src/P2P/MulticastBroadcaster.ts

```diff
@@ -0,0 +1,193 @@
+import KeyManager from '@polykey/KeyManager'
+import RPCMessage from '@polykey/RPC/RPCMessage'
+import dgram from 'dgram'
+import crypto from 'crypto'
+import { EventEmitter } from 'events'
+
+// This module is based heavily on libp2p's mDNS module:
+// https://github.com/libp2p/js-libp2p-mdns
+// It is supposed to discover peers on the local network
+// This module was also generated with the help of:
+// https://nrempel.com/using-udp-multicast-with-node-js/
+//
+// """
+// In computer networking, the multicast DNS (mDNS) protocol
+// resolves hostnames to IP addresses within small networks
+// that do not include a local name server
+// """
+
+const UDP_MULTICAST_PORT = 5353;
+const UDP_MULTICAST_ADDR = "224.0.0.251";
+
+type PeerMessage = {
+  encryptedLocalPubKey: Buffer
+  encryptedPeerPubKey: Buffer
+  rawRandomMessage: Buffer
+  encryptedRandomMessage: Buffer
+}
+
+class MulticastBroadcaster extends EventEmitter {
+  peerStore: PeerStore
+  keyManager: KeyManager
+
+  socket: dgram.Socket
+
+  interval: number
+  queryInterval: NodeJS.Timeout | null
+  peerPubKeyMessages: Map<string, PeerMessage> = new Map()
+  constructor(
+    peerStore: PeerStore,
+    keyManager: KeyManager
+  ) {
+    super()
+
+    this.peerStore = peerStore
+    this.keyManager = keyManager
+
+    this.interval = (1e3)
+    this.queryInterval = null
+  }
+
+
+  queryLAN() {
+    const query = () => {
+      for (const pubKey of this.peerPubKeyMessages.keys()) {
+        const peerMessage = this.peerPubKeyMessages.get(pubKey)
+        if (peerMessage) {
+          const handshakeMessage = RPCMessage.encodeHandshakeMessage(
+            peerMessage.encryptedPeerPubKey,
+            peerMessage.encryptedLocalPubKey,
+            peerMessage.encryptedRandomMessage
+          )
+
+          this.socket.send(handshakeMessage, 0, handshakeMessage.length, UDP_MULTICAST_PORT, UDP_MULTICAST_ADDR, () => {
+            console.info(`Sending message to peer`);
+          });
+        }
+
+      }
+    }
+
+    // Immediately start a query, then do it every interval.
+    query()
+    return setInterval(query, this.interval)
+  }
+
+  /**
+   * Start sending queries to the LAN.
+   *
+   * @returns {void}
+   */
+  async start(): Promise<void> {
+    // Create socket
+    this.socket = dgram.createSocket({ type: "udp4", reuseAddr: true })
+    this.socket.bind(UDP_MULTICAST_PORT)
+
+    // Set up listener
+    this.socket.on("listening", (() => {
+      this.socket.addMembership(UDP_MULTICAST_ADDR);
+      const address = this.socket.address();
+      console.log(
+        `UDP socket listening on ${address.address}:${address.port} pid: ${
+          process.pid
+        }`
+      );
+    }).bind(this));
+
+    // Handle messages
+    this.socket.on("message", this.handleMessages.bind(this));
+
+    // Start the query process
+    this.queryInterval = this.queryLAN()
+  }
+
+  private async handleMessages(message: any, rinfo: any) {
```

Comment:
I'm thinking that this class handles a specific domain of messages. Not just any message at all. Perhaps instead we name this more specifically.

Reply:
I agree, I have renamed it to `handleHandshakeMessages`, this could probably be more specific as well, but I think 'handshake' in the context of polykey will probably mean the same thing if it's multicast peer discovery or some other way to handshake (i.e. they will probably need the same fields for the handshake)